frame/support/procedural/src/pallet/expand/hooks.rs

```diff
@@ -0,0 +1,110 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::pallet::Def;
+use syn::spanned::Spanned;
+
+/// * implement the individual traits using the Hooks trait
+pub fn expand_hooks(def: &mut Def) -> proc_macro2::TokenStream {
+	let frame_support = &def.frame_support;
+	let type_impl_gen = &def.type_impl_generics();
+	let type_use_gen = &def.type_use_generics();
+	let pallet_ident = &def.pallet_struct.pallet;
+	let where_clause = &def.hooks.where_clause;
+	let frame_system = &def.frame_system;
+
+	let hooks_item_span = def.item.content.as_mut()
+		.expect("Checked by def parser").1[def.hooks.index].span();
+
+	quote::quote_spanned!(hooks_item_span =>
```

Comment:
I get what is happening here, but I think it is worth thinking: Given `trait hooks`, do we need the 4 individual traits (OnInitialize, OnFinalize, ..)? If we do, what is the rationale? Both are _**frame**_  terms anyhow, meaning that all frame runtimes have them all, so I see less reason to keep the individual traits.

Reply:
This could very well be noted as an issue for an upcoming PR, if you agree.

Reply:
The main reason was to avoid breaking other crates like frame-executive use those traits, and to avoid changing decl_module.

Then I also used this architecture to avoid modifying input when adding custom logic in OnRuntimeUpgrade:
OnRuntimeUpgrade is implemented using Hook::on_runtime_upgrade + some code written by macro to write the pallet version.

The second point is debatable:
* it avoid having confusion on what is the implementation of `Hooks` as the implementation of `Hooks` is untouched.
* it might bring some confusion in the implementation of `OnRuntimeUpgrade` as some user could expect that is it equal to implementation of `Hooks::on_runtime_upgrade`.

I kind of like having the distinction between Hooks and OnRuntimeUpgrade so that macro added logic doesn't require touching inputs. At the same time I agree this kind of create redundant traits. all in all I see Hooks as a trait to be used to give information to pallet macro in a rust way.